# PIPELINE OUTPUT DOCUMENTATION

The output file tree should look like the tree bellow:

```bash
<output_dir>/
├── <sample_id>
│   ├── <taxid>
│   │   ├── <sample_id>.<taxid>.consensus.fa
│   │   ├── <sample_id>.<taxid>.qc.csv
│   │   ├── <sample_id>.<taxid>.sorted.bam
│   │   └── <sample_id>.<taxid>.sorted.bam.bai
│   ├── [...]
│   ├── <sample_id>.class_seqs_1.fq
│   ├── <sample_id>.class_seqs_2.fq
│   ├── <sample_id>.kraken.output
│   ├── <sample_id>.report.txt
│   ├── <sample_id>.unclass_seqs_1.fq
│   ├── <sample_id>.unclass_seqs_2.fq
├── [...]
├── classification_report.csv
```

## Main output

- **Report CSV**: A csv file summarizing consensus sequences information and qc metrics obtained from each input samples. This file i
  - location: `<output_dir>/classification_report.csv`
- **Consensus sequences per taxid**: the consensus sequence obtained for a given `taxid` and `sample_id` combination.
  - location: `<output_dir/<sample_id>/<sample_id>.<taxid>.consensus.fa`

## Intemediate output files

**Kraken Output files** generated by [run_kraken process](./docs/modules/run_kraken.md)

- `<sample_id>.class_seqs_{1,2}.fq`: The pair of fastq files containing all reads which were associated to a `taxid` in the database.
- `<sample_id>.unclass_seqs_{1,2}.fq`: The pair of fastq files containing all reads which were not classified to any `taxid` in the database
- `<sample_id>.report.txt`: A `tsv` file sumarizing the number of reads associated to a given item in the taxonomic tree of the kraken database. For more details, check [this file format kraken2 documentation](https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#sample-report-output-format)

  - Here is an example of what the content of this file should look like:

```tsv
81.03	3954901	3954901	U	0	unclassified
18.97	926041	0	R	1	root
18.97	926041	29	D	10239	  Viruses
18.76	915890	0	D1	2559587	    Riboviria
18.76	915890	0	K	2732396	      Orthornavirae
17.42	850471	0	P	2732408	        Pisuviricota
17.42	850445	0	C	2732506	          Pisoniviricetes
17.42	850445	0	O	76804	            Nidovirales
17.42	850445	0	O1	2499399	              Cornidovirineae
17.42	850445	0	F	11118	                Coronaviridae
[...]
```

- `<sample_id>.kraken.output`: a `tsv` file containing classification of each reads id. For more details, check [this kraken2 output file specific documentation](https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#standard-kraken-output-format)
  - here is an example of its content:

```
U	A00948:701:HJHWGDSX7:2:1101:4218:1031	0	151|151	A:1 0:116 |:| 0:117
U	A00948:701:HJHWGDSX7:2:1101:9498:1031	0	151|151	A:1 0:116 |:| 0:117
U	A00948:701:HJHWGDSX7:2:1101:7699:1047	0	151|151	A:1 0:116 |:| 0:117
C	A00948:701:HJHWGDSX7:2:1101:15058:1047	2697049	151|151	A:1 2697049:116 |:| 2697049:117
U	A00948:701:HJHWGDSX7:2:1101:30481:1047	0	151|151	A:1 0:116 |:| 0:117
```

The **BWA output BAM files** generated by the [bwa_alignment_and_post_processing process](./modules/bwa_alignment.md)

- `<sample_id>.<taxid>.sorted.bam`: A [Binary Alignment Map (BAM)](https://en.wikipedia.org/wiki/Binary_Alignment_Map) file resulted from the alignment between the `<sample_id>.<taxid>.fastq` pair outputed by kraken2ref output (`run_kraken2ref_dump_fastq`), more details [here](./modules/run_kraken2ref_and_pre_report.md).
- `<sample_id>.<taxid>.sorted.bai`: the index file of the `sorted.bam` file. the BAI file provide fast retrieval of alignments overlapping a specified region without going through the whole alignments, check [section 5 of the samtools documenation](https://samtools.github.io/hts-specs/SAMv1.pdf) for a more detailed description.

The **QC report** which is the output of `run_qc_script`, check [here](./modules/run_qc_script.md) for more detail.

- `<sample_id>.<taxid>.qc.csv`: a csv containing QC metrics of the consensus sequence obtained (`<sample_id.<taxid>.consensus.fa`).

```csv
sample_name,pct_N_bases,pct_covered_bases,longest_no_N_run,num_aligned_reads,fasta,bam,qc_pass,total_mapped_reads,ivar_md,total_unmapped_reads
<sample_id.<taxid>,0.76,99.19,3543,129724,<sample_id.<taxid>.consensus.fa,<sample_id.<taxid>.sorted.bam,TRUE,129837,10,0
```

